app/main.py

```diff
@@ -0,0 +1,97 @@
+import bcrypt,jwt
+from fastapi import FastAPI, status
+from fastapi.exceptions import HTTPException
+from sqlmodel import Session, select
+from starlette.middleware.cors import CORSMiddleware
+
+from db import engine, create_db_and_tables, SECRET_KEY, ALGORITHM
+from models.models import User, UserUpdate, Token, LoginRequest
+
+app = FastAPI()
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+@app.post("/users/signup", status_code=status.HTTP_201_CREATED)
+def create_user(user:User):
+    with Session(engine) as session:
+        if session.exec(select(User).filter(User.email == user.email)).first():
+            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail = "user is exist")
+
+        user.password = bcrypt.hashpw(user.password.encode('utf-8'), bcrypt.gensalt()).decode() 
+        session.add(user)
+        session.commit()
+        session.close()
+        return {"message" : "create user"}
+ 
+@app.post("/users/login", response_model=Token)
+async def login_for_access_token(login_request: LoginRequest):
+    user = authenticate_user(email=login_request.email, password=login_request.password)
+    if not user:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Incorrect email or password",
+            headers={"WWW-Authenticate": "Bearer"},
+        )
+   
+    access_token = create_access_token(email=user.email)
+
+    return {"access_token": access_token, "token_type": "bearer"}
+
+def authenticate_user(email: str, password: str):
+    with Session(engine) as session:
+        user = session.exec(select(User).filter(User.email == email)).first()
+
+    if not user:
+        return None
+    if not check_password(password, user.password):
+        return None
+    return user
+
+def create_access_token(email: str):
+    payload = {"email": email}
+    encoded_jwt = jwt.encode(payload, SECRET_KEY, algorithm=ALGORITHM)
+    return encoded_jwt
+
+def check_password(password:str, db_password:str):
+    checkpw = bcrypt.checkpw(password.encode('utf-8'), db_password.encode('utf-8'))
+    return checkpw 
```

Comment:
P2: 외부로 export 되지 않는 함수들 앞에는 underscore(_)를 붙여주세요. 또한 타입 힌트도 같이 추가해주시면 감사드릴께요.

```suggestion
def _authenticate_user(email: str, password: str) -> Optional[User]:
    with Session(engine) as session:
        user = session.exec(select(User).filter(User.email == email)).first()

    if not user:
        return None
    if not check_password(password, user.password):
        return None
    return user

def _create_access_token(email: str) -> str:
    payload = {"email": email}
    return = jwt.encode(payload, SECRET_KEY, algorithm=ALGORITHM)

def _check_password(password:str, db_password:str) -> bool:
    return bcrypt.checkpw(password.encode('utf-8'), db_password.encode('utf-8'))
```

Reply:
네! 수정하였습니다!